### ISSUE TITLE
Support webserver caching: set ETag and Last-Modified headers on response, compare If-Modified-Since header on request, return 304 Not Modified if not stale

### ISSUE THREAD

**Issue by tobscure** [flarum/core#980]
See:
http://whitequark.org/blog/2014/04/05/page-caching-with-nginx/
http://apidock.com/rails/ActionController/ConditionalGet/stale%3F

**Comment by webeindustry** [comment]
After running into this issue today:
#985 

I'm permanently disabling Nginx cache in favor of Varnish for the API System. 

Viewing the [api docs](http://flarum.org/docs/api/) it appears the best approach is to start by issuing a http purge request for every unsafe API call. 

So for example:

```
POST /api/discussions - create a new discussion
PATCH /api/discussions/:id - update a discussion
DELETE /api/discussions/:id - delete a discussion

```
Would need generate a http purge upstream to varnish. This would be filtered through the VCL and expire the `GET /api/discussions/:id - get a discussion by ID`

It's a starting point, but what to do with objects that are shared outside this request? The next step appears to be utilizing the BAN function though to be entirely honest I'm not sure exactly how this would work. 

See [here](https://www.smashingmagazine.com/2014/04/cache-invalidation-strategies-with-varnish-cache/) for suggestions.

The gist is that there's at least a couple ways to go about getting varnish to know which objects in a request are shared and on which other pages. This gets fed back to a regex ban rule. 

Going this route is more complicated than nginx, but more specific, and the safe way to cache dynamic content which the request contains a mix of public/private data.

**Comment by Luceos** [comment]
@webeindustry was this issue created to support caching on the hosting services?

**Comment by BartVB** [comment]
Would be great if this was implemented in a cache agnostic way. I.e. don't depend on custom VCL in Varnish but use proper PURGE requests to update the cache(s) where needed and make sure that you can send a 304 really quickly.

**Comment by dav-is** [comment]
Caching in Flarum is very difficult to implement if not impossible because of the way the api data is structured. Each api response includes user data about permissions, etc. We tried and failed.

**Comment by BartVB** [comment]
Even for guests?
Guest visits are easiest to cache and also the most effective.

**Comment by webeindustry** [comment]
Yes, even for guests. Have at it. It's a time waster as it's currently coded.

**Comment by tobscure** [comment]
@dav-is @webeindustry what are the specific problems prevent guest caching from being feasible?

**Comment by webeindustry** [comment]
Sorry I've been out of town it seems my memory is off on this one that was a while ago now. @dav-is told me we didn't really have an issue with guest cache, but got so bummed that logged in sessions were a fail we scraped the whole idea until a better solution was implemented.

**Comment by dav-is** [comment]
The issue being permissions and read status are sent with every logged in api Request

**Comment by BartVB** [comment]
> The issue being permissions and read status are sent with every logged in api Request

This shouldn't be a problem for requests from guests?

I think this issue should be split into several parts:
- Send proper 304s (should be possible for both guests and logged in users)
- Send headers to make implementing a reverse proxy (or other caching) feasible (for example `Cache-Control: public`
- See if it's possible/feasible to cache data for registered users
